test/IDE/complete_tf_194.swift

```diff
@@ -0,0 +1,7 @@
+// https://bugs.swift.org/browse/TF-194: invalid unary ops crash repl completer
+
+// The ASTVerifier doesn't like this AST.
+// XFAIL: swift_ast_verifier
```

Comment:
I think this is worth digging into. What doesn't it like about it? If the rule is "every expression should have a type, even if it's an error type, after type-checking", maybe that's a better fix: find the place where we fail to give the expression a type due to the unknown operator and adjust it there.

(For comparison, what do we do for unresolved identifiers? Those have to be just as type-less, right?)

Reply:
The ASTVerifier error from this testcase is:
```
expression has no type
(error_expr implicit type='<null>')Stack dump:
0.	Program arguments: /usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test -target x86_64-unknown-linux-gnu -module-cache-path /usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/clang-module-cache -completion-cache-path /usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/completion-cache -swift-version 4 -repl-code-completion -source-filename=/usr/local/google/home/marcrasi/swift-base-master/swift/test/IDE/complete_tf_194.swift 
1.	While walking into decl declaration 0x6d89238 (at <REPL Input>:7:1)
#0 0x0000000003f5bd04 PrintStackTraceSignalHandler(void*) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x3f5bd04)
#1 0x0000000003f59e60 llvm::sys::RunSignalHandlers() (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x3f59e60)
#2 0x0000000003f5beb2 SignalHandler(int) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x3f5beb2)
#3 0x00007f414109d0c0 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x110c0)
#4 0x00007f413f465fcf gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x32fcf)
#5 0x00007f413f4673fa abort (/lib/x86_64-linux-gnu/libc.so.6+0x343fa)
#6 0x00000000010a87a7 (anonymous namespace)::Verifier::walkToExprPost(swift::Expr*) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x10a87a7)
#7 0x00000000010bb777 (anonymous namespace)::Traversal::visitApplyExpr(swift::ApplyExpr*) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x10bb777)
#8 0x00000000010bbb78 swift::ASTVisitor<(anonymous namespace)::Traversal, swift::Expr*, swift::Stmt*, bool, swift::Pattern*, bool, void>::visit(swift::Stmt*) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x10bbb78)
#9 0x00000000010bdaf7 swift::ASTVisitor<(anonymous namespace)::Traversal, swift::Expr*, swift::Stmt*, bool, swift::Pattern*, bool, void>::visit(swift::Decl*) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x10bdaf7)
#10 0x00000000010b814c (anonymous namespace)::Traversal::doIt(swift::Decl*) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x10b814c)
#11 0x00000000010b8073 swift::Decl::walk(swift::ASTWalker&) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x10b8073)
#12 0x000000000116581e swift::SourceFile::walk(swift::ASTWalker&) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x116581e)
#13 0x00000000010a2e93 swift::verify(swift::SourceFile&) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x10a2e93)
#14 0x0000000000d4dd60 swift::performTypeChecking(swift::SourceFile&, swift::TopLevelContext&, swift::OptionSet<swift::TypeCheckingFlags, unsigned int>, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0xd4dd60)
#15 0x000000000058beda doCodeCompletion(swift::SourceFile&, llvm::StringRef, unsigned int*, swift::CodeCompletionCallbacksFactory*) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x58beda)
#16 0x000000000058b960 swift::REPLCompletions::populate(swift::SourceFile&, llvm::StringRef) (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x58b960)
#17 0x0000000000496bff main (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x496bff)
#18 0x00007f413f4532b1 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202b1)
#19 0x000000000049148a _start (/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test+0x49148a)
/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/test-linux-x86_64/IDE/Output/complete_tf_194.swift.script: line 1: 204289 Aborted                 /usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/bin/swift-ide-test -target x86_64-unknown-linux-gnu -module-cache-path '/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/clang-module-cache' -completion-cache-path '/usr/local/google/home/marcrasi/swift-base-master/build/buildbot_linux/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/completion-cache' -swift-version 4 -repl-code-completion -source-filename=/usr/local/google/home/marcrasi/swift-base-master/swift/test/IDE/complete_tf_194.swift
```

By the way, I'm having a very similar discussion about ASTVerifier+codecompletion in https://github.com/apple/swift/pull/22580. It might turn out that we shouldn't run the ASTVerifier whenever the code-completion produces an AST with errors. If that happens, then this ASTVerifier problem will go away.

I tried a few examples with unresolved identifiers and it seems to always generate ASTs that pass the verifier:
```
Input to %target-swift-ide-test -repl-code-completion:
unresolved
AST:
(source_file ""
  (top_level_code_decl range=[<REPL Input>:9:1 - line:9:1]
    (brace_stmt range=[<REPL Input>:9:1 - line:9:1])))
```
```
Input to %target-swift-ide-test -repl-code-completion:
func foo() { return bar }; fo
AST:
(source_file ""
  (func_decl range=[<REPL Input>:9:1 - line:9:25] trailing_semi "foo()" interface type='() -> ()' access=internal
    (parameter_list range=[<REPL Input>:9:9 - line:9:10]))
  (top_level_code_decl range=[<REPL Input>:9:28 - line:9:28]
    (brace_stmt range=[<REPL Input>:9:28 - line:9:28])))
```
Not sure why the unresolved identifiers aren't showing up in these ASTs.

Reply:
The verifier already distinguishes "parsed" from "type-checked but has errors" and "type-checked without errors". I'm not sure it's worth adding a distinction of "type-checked but only some stuff", but that might in fact be the right behavior based on the code completion point. We'd still want to run the "parsed" verifier at least, though, and ideally we'd still do verification on all the *other* declarations that *don't* contain the code completion point.

Reply:
It seems that preventing the ASTVerifier from running when there are errors is the right thing to do. I created a PR that does that: #22591

That actually fixes this crasher too, so I'll just close this PR.